snuba/query/processors/mapping_promoter.py

```diff
@@ -0,0 +1,142 @@
+from typing import Mapping, NamedTuple, Optional
+
+from snuba.clickhouse.columns import ColumnSet
+from snuba.clickhouse.processors import QueryProcessor
+from snuba.clickhouse.query import Query
+from snuba.query.expressions import Column, Expression, FunctionCall, Literal
+from snuba.request.request_settings import RequestSettings
+
+
+class SubscriptableMatch(NamedTuple):
+    # The table name associated with the nested column found in the query.
+    table_name: Optional[str]
+    # The nested column name
+    column_name: str
+    # The key column name in the mapping (like key for tags.key)
+    key_field: str
+    # The value column name in the mapping (like key for tags.value)
+    val_field: str
+    # The key found in the query (like tags[key])
+    key: str
+
+
+def match_subscriptable_reference(exp: Expression) -> Optional[SubscriptableMatch]:
```

Comment:
I expect to reuse this logic for other tags optimizations like flattened tags. Anyway it seems to me more readable if it is not mixed into the process_query logic